doc/sphinx/proof-engine/ltac.rst

```diff
@@ -713,339 +962,563 @@ tactic
 
-   .. example::
+         match True with
+         | True => idtac "branch 1"; fail
+         | _ => idtac "branch 2"
+         end.
 
-      .. coqtop:: all abort
+.. _match_vs_multimatch_ex:
 
-         Ltac f x :=
-           match x with
-             context f [S ?X] =>
-             idtac X;                    (* To display the evaluation order *)
-             assert (p := eq_refl 1 : X=1);    (* To filter the case X=1 *)
-             let x:= context f[O] in assert (x=O) (* To observe the context *)
+   .. example:: Comparison of match and multimatch
+
+      :tacn:`match` tactics are only evaluated once, whereas :tacn:`multimatch`
+      tactics may be evaluated more than once if the following constructs trigger backtracking:
+
+      .. coqtop:: all fail
+
+         Fail match True with
+         | True => idtac "branch 1"
+         | _ => idtac "branch 2"
+         end ;
+         match True with
+         | True => idtac "branch A"; fail
+         end.
+
+      .. coqtop:: all fail
+
+         Fail multimatch True with
+         | True => idtac "branch 1"
+         | _ => idtac "branch 2"
+         end ;
+         match True with
+         | True => idtac "branch A"; fail
+         end.
+
+.. _match_with_holes_ex:
+
+   .. example:: Matching a pattern with holes
+
+      Notice the :tacn:`idtac` prints ``(z + 1)`` while the :tacn:`pose` substitutes
+      ``(x + 1)``.
+
+      .. coqtop:: in reset
+
+         Goal True.
+
+      .. coqtop:: all
+
+           match constr:(fun x => (x + 1) * 3) with
+           | fun z => ?y * 3 => idtac "y =" y; pose (fun z: nat => y * 5)
            end.
+
+.. _match_term_context_ex:
+
+   .. example:: Multiple matches for a "context" pattern.
+
+      Internally "x <> y" is represented as "(not x y)", which produces the
+      first match.
+
+      .. coqtop:: in reset
+
+         Ltac f t := match t with
+                    | context [ (not ?t) ] => idtac "?t = " t; fail
+                    | _ => idtac
+                    end.
          Goal True.
-         f (3+4).
+
+      .. coqtop:: all
+
+         f ((not True) <> (not False)).
 
 .. _ltac-match-goal:
 
-Pattern matching on goals
-~~~~~~~~~~~~~~~~~~~~~~~~~
+Pattern matching on goals and hypotheses: match goal
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-We can perform pattern matching on goals using the following expression:
+.. tacn:: @match_key {? reverse } goal with {? %| } {+| @goal_pattern => @ltac_expr } end
+   :name: lazymatch goal; match goal; multimatch goal
 
-.. we should provide the full grammar here
+   .. insertprodn goal_pattern match_hyp
 
-.. tacn:: match goal with {+| {+, @context_hyp} |- @cpattern => @ltac_expr } | _ => @ltac_expr end
-   :name: match goal
+   .. prodn::
+      goal_pattern ::= {*, @match_hyp } |- @match_pattern
+      | [ {*, @match_hyp } |- @match_pattern ]
+      | _
+      match_hyp ::= @name : @match_pattern
+      | @name := {? [ @match_pattern ] : } @match_pattern
```

Comment:
`match_hyp` is not explained but quite intuitive to understand, except the square brackets (which may simply be there so that the parser can figure out what is part of the term and what is part of the type) and the way the `{? }` is wrapped. I wonder if it wouldn't be clearer if it was split into two rules:

```
      match_hyp ::= @name : @match_pattern
      | @name := @match_pattern
      | @name := [ @match_pattern ] : @match_pattern
```

Reply:
>  except the square brackets (which may simply be there so that the parser can figure out what is part of the term and what is part of the type)

Note that the square brackets are only needed if you overwrite the infix `:` notation with your own notation.  Otherwise `@name := @match_pattern : @match_pattern` will be exactly the same as `@name := [ @match_pattern ] : @match_pattern`  Note also that the square brackets mean that if you want to match a context-bound list, you either have to double the brackets or wrap your list in parentheses.  Note also that when `@match_pattern` is `[?x ; ?y]` (for list), the `@name := @match_pattern` fails to parse correctly, and instead there is an error about a missing `:`

Reply:
That's a lot of interesting remarks probably worth writing down in the doc...

Reply:
These are not intuitive to me.  Would you explain?

```
      | @name := @match_pattern
      | @name := [ @match_pattern ] : @match_pattern
```

Reply:
Consider the goal in
```coq
Goal let x := True in let y := 0 in False -> nat -> True.
  intros.
  (* 1 subgoal (ID 6)

  x := True : Prop
  y := 0 : nat
  H : False
  H0 : nat
  ============================
  True
*)
```
The pattern `v := ?X` will match `x` (with `X` bound to `True`) and `y` (with `X` bound to `0`).  The pattern `v := [ ?X ] : ?T` and `v := ?X : ?T` (which are identical unless you've overridden the infix `:` notation) will match `x` (with `X` bound to `True` and `T` bound to `Prop`) and `y` (with `X` bound to `0` and `T` bound to `nat`).  The pattern `v : ?T` will match any of the hypotheses, both the ones with bodies and the ones without.

Here are two useful tactics involving this sort of thing:
```coq
Ltac subst_local :=
  repeat match goal with
         | [ v := _ |- _ ] => subst v
         end.
Ltac subst_evars :=
  repeat match goal with
         | [ v := ?e |- _ ] => is_evar e; subst v
         end.
```

Reply:
> `@name := @match_pattern : @match_pattern`

That's not the underlying syntax for `v := ?X : ?T`, which is actually `@name := @term10 : @term`.

The description for that `term` is "The expression `term10 : type` is a type cast expression. It enforces the type of `term10` to be `type`.".  When we have a hypothesis "y := 0 : nat", we're not coercing it.  How we word that description better?

Reply:
While there is this notation in `ssreflect.v` associated with `:`, there is no equivalent default notation, right?

`Notation "x : T" := (CoqCast x T) : core_scope.`

Reply:
> Note also that when @match_pattern is [?x ; ?y] (for list), the @name := @match_pattern fails to parse correctly, and instead there is an error about a missing :

The workaround is to use `(cons ?x ?y)` instead?  Is this a fixable bug or something we have to live with?  Generally I lean toward not documenting fixable bugs.

Reply:
How does this look?

![image](https://user-images.githubusercontent.com/1253341/83934317-e9d77080-a764-11ea-8f16-c3358541b1ef.png)

Reply:
> > `@name := @match_pattern : @match_pattern`
> 
> That's not the underlying syntax for `v := ?X : ?T`, which is actually `@name := @term10 : @term`.
> 
> The description for that `term` is "The expression `term10 : type` is a type cast expression. It enforces the type of `term10` to be `type`.". When we have a hypothesis "y := 0 : nat", we're not coercing it. How we word that description better?

I believe what happens is that the parsing code explicitly strips off outermost cast nodes and uses them to match the type instead.  See https://github.com/coq/coq/blob/9c26e583668827bff5159e7671c406ace8b2f3ae/plugins/ltac/g_ltac.mlg#L243-L250

> While there is this notation in `ssreflect.v` associated with `:`, there is no equivalent default notation, right?
> 
> `Notation "x : T" := (CoqCast x T) : core_scope.`

That's correct

Reply:
> How does this look?

Looks good to me!

Reply:
Though you might want to explicitly mention that `:` is overridden if you `Require Import ssreflect`

Reply:
> The workaround is to use `(cons ?x ?y)` instead? Or `([?x ; ?y])`? Is this a fixable bug or something we have to live with? Generally I lean toward not documenting fixable bugs.

Yes, both of those work.  I don't know how to fix it, I suspect it requires knowing the parser gods much more deeply than I do.  I suspect it in fact requires multi-success parsing, and is thus not fixable, but I'm not sure.  @roglo or @ppedrot or @herbelin might whether or not (and how) to factorize the Ltac parser (or might know who knows) so that we can get both matches on https://github.com/coq/coq/blob/9c26e583668827bff5159e7671c406ace8b2f3ae/plugins/ltac/g_ltac.mlg#L242 and similar-looking matches with list notations in https://github.com/coq/coq/blob/9c26e583668827bff5159e7671c406ace8b2f3ae/plugins/ltac/g_ltac.mlg#L243 (i.e., so that `x := [?y ; ?z] : ?T` and `x := [?y ; ?z]` match the second rule), but I certainly don't.  And I don't see even in principle how this could be fixed for the singleton list notation.  So I'm inclined to think it's not really fixable.

Reply:
> I suspect it in fact requires multi-success parsing

If it's not ambiguous it should be possible to handle it.  One useful trick is to look ahead one or more tokens before making a parse decision.  Camlp5 b y default only looks at the next token.  If you can describe in words how to distinguish them, email me and I'll give a little thought.  Not in this PR, please!  Maybe just submit an issue and cc me (since I don't routinely read new issues).

multisuccess parsing, hmm, sounds dreadful--adds a lot of complexity.  If a multimatch fails, would we backtrack into the parser :-) ?